### PR TITLE
Complete Actors, Devices and Groups UIs

### PR DESCRIPTION
**Pull request by AndrewDryga**
This will be done once the remaining UI code is covered with tests.

### CODE REVIEW

File: elixir/apps/api/lib/api/device/channel.ex [hunk 1]
```diff
@@ -85,6 +85,7 @@ defmodule API.Device.Channel do
     connected_gateway_ids = Map.get(attrs, "connected_gateway_ids", [])
 
     with {:ok, resource} <- Resources.fetch_resource_by_id(resource_id, socket.assigns.subject),
+         # TODO:
          # :ok = Resource.authorize(resource, socket.assigns.subject),
```
Comment (AndrewDryga) [inline elixir/apps/api/lib/api/device/channel.ex:89]: @bmanifold if you feel like it - you can also handle policy evaluation when connecting to a resource.

File: elixir/apps/api/test/support/channel_case.ex [hunk 2]
```diff
@@ -14,6 +14,7 @@ defmodule API.ChannelCase do
       import Phoenix.ChannelTest
       import API.ChannelCase
       alias Domain.Repo
+      alias Domain.Fixtures
```
Comment (AndrewDryga) [inline elixir/apps/api/test/support/channel_case.ex:17]: @bmanifold I streamlined fixture naming so there is no need to `alias` a ton of fixtures/factories in our tests, we can just use `Fixtures.NameOfDomainContext` anywhere.

File: elixir/apps/domain/lib/domain/actors.ex [hunk 3]
```diff
@@ -1,68 +1,235 @@
...
-    |> Enum.reduce(Ecto.Multi.new(), fn {provider_identifier, attrs}, multi ->
-      Ecto.Multi.insert(
-        multi,
-        {:group, provider_identifier},
-        Group.Changeset.create_changeset(provider, provider_identifier, attrs),
-        conflict_target: Group.Changeset.upsert_conflict_target(),
-        on_conflict: Group.Changeset.upsert_on_conflict(),
-        returning: true
-      )
+  def sync_provider_groups_multi(%Auth.Provider{} = provider, attrs_list) do
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/actors.ex]: :trollface: multi

File: elixir/apps/domain/lib/domain/actors.ex [hunk 4]
```diff
@@ -163,52 +323,42 @@ defmodule Domain.Actors do
...
 
-  def create_actor(%Auth.Provider{} = provider, provider_identifier, attrs) do
-    {provider_attrs, attrs} = Map.pop(attrs, "provider", %{})
-
+  def create_actor(%Auth.Provider{} = provider, attrs) do
     Ecto.Multi.new()
     |> Ecto.Multi.insert(:actor, Actor.Changeset.create_changeset(provider.account_id, attrs))
     |> Ecto.Multi.run(:identity, fn _repo, %{actor: actor} ->
-      Auth.create_identity(actor, provider, provider_identifier, provider_attrs)
+      # TODO: we should not reuse the same attrs here, instead we should use attrs["identities"] list for that,
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/actors.ex]: TODO

File: elixir/apps/domain/lib/domain/actors.ex [hunk 5]
```diff
@@ -76,6 +243,7 @@ defmodule Domain.Actors do
   end
 
   def change_group(%Group{}, _attrs) do
+    # TODO: we can change but only name
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/actors.ex]: TODO

File: elixir/apps/domain/lib/domain/actors.ex [hunk 6]
```diff
@@ -286,6 +447,30 @@ defmodule Domain.Actors do
     end
   end
 
+  # TODO: when actor is synced we should not allow changing the name
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/actors.ex:288]: TODO

File: elixir/apps/domain/lib/domain/actors/actor/changeset.ex [hunk 7]
```diff
@@ -2,17 +2,29 @@ defmodule Domain.Actors.Actor.Changeset do
   use Domain, :changeset
   alias Domain.Actors
 
-  def create_changeset(account_id, attrs) do
+  # TODO: refactor naming for changeset functions
+  # TODO: defp
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/actors/actor/changeset.ex]: TODO

File: elixir/apps/domain/lib/domain/actors/group/query.ex [hunk 8]
```diff
@@ -32,6 +44,57 @@ defmodule Domain.Actors.Group.Query do
     })
   end
 
+  def preload_few_actors_for_each_group(queryable \\ all(), limit) do
+    queryable
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/actors/group/query.ex:48]: Thanks to @RhodiumToad from #postgresql IRC.

File: elixir/apps/domain/lib/domain/auth.ex [hunk 9]
```diff
@@ -104,6 +128,14 @@ defmodule Domain.Auth do
     |> Repo.list()
   end
 
+  # TODO: and_provisioner == :custom
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/auth.ex]: TODO

File: elixir/apps/domain/lib/domain/auth.ex [hunk 10]
```diff
@@ -398,8 +515,15 @@ defmodule Domain.Auth do
   end
 
   defp build_subject_expires_at(%Actors.Actor{} = actor, expires_at) do
+    now = DateTime.utc_now()
+
     default_session_duration_hours = Map.fetch!(@default_session_duration_hours, actor.type)
-    expires_at || DateTime.utc_now() |> DateTime.add(default_session_duration_hours, :hour)
+    expires_at = expires_at || DateTime.add(now, default_session_duration_hours, :hour)
+
+    max_session_duration_hours = Map.fetch!(@max_session_duration_hours, actor.type)
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/auth.ex:483]: TODO: increase limits and split tokens to allow non-admin-ui tokens to be long lasting for admins

File: elixir/apps/domain/lib/domain/auth/adapters/google_workspace/jobs.ex [hunk 11]
```diff
@@ -1,9 +1,106 @@
 defmodule Domain.Auth.Adapters.GoogleWorkspace.Jobs do
   use Domain.Jobs.Recurrent, otp_app: :domain
+  alias Domain.{Auth, Actors}
+  alias Domain.Auth.Adapters.GoogleWorkspace
   require Logger
 
   every minutes(5), :refresh_access_tokens do
-    Logger.debug("Refreshing tokens")
+    # Enum.each(fn provider ->
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/auth/adapters/google_workspace/jobs.ex]: TODO

File: elixir/apps/domain/lib/domain/auth/identity.ex [hunk 12]
```diff
@@ -18,6 +18,11 @@ defmodule Domain.Auth.Identity do
     field :created_by, Ecto.Enum, values: ~w[system provider identity]a
     belongs_to :created_by_identity, Domain.Auth.Identity
 
+    has_many :devices, Domain.Devices.Device, where: [deleted_at: nil]
+
+    # TODO: we don't want to have deleted_at for actors and identities, instead we should just disable identities
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/auth/identity.ex]: TODO

We also should delete related entries when delete some entity that owns them.

File: elixir/apps/domain/lib/domain/devices.ex [hunk 13]
```diff
@@ -93,15 +106,38 @@ defmodule Domain.Devices do
...
+        |> preload_online_statuses()
+
+      {:ok, devices}
     else
       false -> {:error, :not_found}
       other -> other
     end
   end
 
+  # TODO: this is ugly!
```
Comment (AndrewDryga) [inline elixir/apps/domain/lib/domain/devices.ex:125]: TODO but probably for a different PR

File: elixir/apps/domain/priv/repo/seeds.exs [hunk 14]
```diff
@@ -53,7 +53,7 @@ IO.puts("")
     adapter_config: %{}
   })
 
-{:ok, _oidc_provider} =
+{:ok, oidc_provider} =
   Auth.create_provider(account, %{
```
Comment (AndrewDryga) [inline elixir/apps/domain/priv/repo/seeds.exs:57]: TODO: try to provision a real vault provider here.

File: elixir/apps/domain/test/domain/auth/adapters/google_workspace/jobs_test.exs [hunk 15]
```diff
@@ -0,0 +1,166 @@
+defmodule Domain.Auth.Adapters.GoogleWorkspace.JobsTest do
+  use Domain.DataCase, async: true
+  alias Domain.{Auth, Actors}
+  alias Domain.Mocks.GoogleWorkspaceDirectory
+  import Domain.Auth.Adapters.GoogleWorkspace.Jobs
+
+  describe "sync_directory/1" do
```
Comment (AndrewDryga) [inline elixir/apps/domain/test/domain/auth/adapters/google_workspace/jobs_test.exs:84]: TODO: needs more test cases

File: elixir/apps/domain/lib/domain/devices.ex (hunk 13 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/devices.ex:125]: Moved to separate issue

File: elixir/apps/api/test/support/channel_case.ex (hunk 2 above)
Comment (bmanifold) [reply elixir/apps/api/test/support/channel_case.ex:17]: 👍

**Review by bmanifold — APPROVED** [review]
LGTM

**Review by jamilbk — APPROVED** [review]
📦💨

File: elixir/apps/domain/lib/domain/auth.ex (hunk 10 above)
Comment (AndrewDryga) [reply elixir/apps/domain/lib/domain/auth.ex:483]: Moved to a separate issue

**Pull request merged**